internal/docstore/dynamodocstore/codec.go

```diff
@@ -0,0 +1,316 @@
+// Copyright 2019 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dynamodocstore
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strconv"
+
+	dyn "github.com/aws/aws-sdk-go/service/dynamodb"
+	"gocloud.dev/internal/docstore/driver"
+)
+
+var nullValue = new(dyn.AttributeValue).SetNULL(true)
+
+type encoder struct {
+	av *dyn.AttributeValue
+}
+
+func (e *encoder) EncodeNil()            { e.av = nullValue }
+func (e *encoder) EncodeBool(x bool)     { e.av = new(dyn.AttributeValue).SetBOOL(x) }
+func (e *encoder) EncodeInt(x int64)     { e.av = new(dyn.AttributeValue).SetN(strconv.FormatInt(x, 10)) }
+func (e *encoder) EncodeUint(x uint64)   { e.av = new(dyn.AttributeValue).SetN(strconv.FormatUint(x, 10)) }
+func (e *encoder) EncodeBytes(x []byte)  { e.av = new(dyn.AttributeValue).SetB(x) }
+func (e *encoder) EncodeFloat(x float64) { e.av = encodeFloat(x) }
+
+func (e *encoder) ListIndex(int) { panic("impossible") }
+func (e *encoder) MapKey(string) { panic("impossible") }
+
+func (e *encoder) EncodeString(x string) {
+	if len(x) == 0 {
+		e.av = nullValue
+	} else {
+		e.av = new(dyn.AttributeValue).SetS(x)
+	}
+}
+
+func (e *encoder) EncodeComplex(x complex128) {
+	e.av = new(dyn.AttributeValue).SetL([]*dyn.AttributeValue{encodeFloat(real(x)), encodeFloat(imag(x))})
+}
+
+func (e *encoder) EncodeList(n int) driver.Encoder {
+	s := make([]*dyn.AttributeValue, n)
+	e.av = new(dyn.AttributeValue).SetL(s)
+	return &listEncoder{s: s}
+}
+
+func (e *encoder) EncodeMap(n int) driver.Encoder {
+	m := make(map[string]*dyn.AttributeValue, n)
+	e.av = new(dyn.AttributeValue).SetM(m)
+	return &mapEncoder{m: m}
+}
+
+func (*encoder) EncodeSpecial(reflect.Value) (bool, error) {
+	return false, nil
+}
+
+type listEncoder struct {
+	s []*dyn.AttributeValue
+	encoder
+}
+
+func (e *listEncoder) ListIndex(i int) { e.s[i] = e.av }
+
+type mapEncoder struct {
+	m map[string]*dyn.AttributeValue
+	encoder
+}
+
+func (e *mapEncoder) MapKey(k string) { e.m[k] = e.av }
+
+func encodeDoc(doc driver.Document) (map[string]*dyn.AttributeValue, error) {
+	var e encoder
+	if err := doc.Encode(&e); err != nil {
+		return nil, err
+	}
+	return e.av.M, nil
+}
+
+// TODO: improve driver.Encoder to make this efficient.
+func encodeDocKeyFields(doc driver.Document, pkey, skey string) (map[string]*dyn.AttributeValue, error) {
```

Comment:
Could you add some documentation to this function? It was not super obvious what's going on here, expecially what the arguments are.

Reply:
Done in response to earlier comment.